### PR TITLE
Data source for Service API

### PR DESCRIPTION
**Pull request by msg4sunny**
Adding data source for Service API

### CODE REVIEW

File: opsgenie/data_source_opsgenie_service.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	client, err := service.NewClient(meta.(*OpsgenieClient).client.Config)
+	if err != nil {
+		return err
+	}
+	name := d.Get("name").(string)
+
+	log.Printf("[INFO] Reading OpsGenie service '%s'", name)
+
+	res, err := client.Get(context.Background(), &service.GetRequest{
+		Id: d.Id(),
```
Comment (ffahri) [inline opsgenie/data_source_opsgenie_service.go]: We shouldn't need to put ID here, users only give us service name and team name, etc

File: opsgenie/data_source_opsgenie_service_test.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+			}
+		}
+
+		return nil
+	}
+}
+
+func testAccDataSourceOpsGenieServiceConfig(randomUserName, randomTeamName, randomServiceName string) string {
+	return fmt.Sprintf(`
+resource "opsgenie_user" "test" {
```
Comment (ffahri) [inline opsgenie/data_source_opsgenie_service_test.go]: why do we need users?

File: opsgenie/data_source_opsgenie_service_test.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  username  = "genietest-%s@opsgenie.com"
+  full_name = "Acceptance Test User"
+  role      = "User"
+}
+resource "opsgenie_user" "test2" {
```
Comment (ffahri) [inline opsgenie/data_source_opsgenie_service_test.go]: same as above

File: opsgenie/data_source_opsgenie_service.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+		},
+	}
+}
+
+func dataSourceOpsGenieServiceRead(d *schema.ResourceData, meta interface{}) error {
+	client, err := service.NewClient(meta.(*OpsgenieClient).client.Config)
+	if err != nil {
+		return err
+	}
+	name := d.Get("name").(string)
```
Comment (ffahri) [inline opsgenie/data_source_opsgenie_service.go:41]: we don't use this variable?

File: opsgenie/data_source_opsgenie_service_test.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    role     = "admin"
+  }
+}
+resource "opsgenie_service" "test" {
+  name = "genieservice-%s"
+  description = "This is our main service"
+  team_id = "${opsgenie_team.test.id}"
+}
+data "opsgenie_service" "existingservice" {
+  name = "${opsgenie_service.test.name}"
```
Comment (ffahri) [inline opsgenie/data_source_opsgenie_service_test.go:66]: we define the only name but our data resource will only know how to read via ID, therefore tests won't pass.
also, you set team_id is required but you didn't define here

```
------- Stdout: -------
=== RUN   TestAccDataSourceOpsGenieService_Basic
--- FAIL: TestAccDataSourceOpsGenieService_Basic (0.08s)
testing.go:569: Step 0 error: config is invalid: Missing required argument: The argument "team_id" is required, but no definition was found.
FAIL
```

**Review by ffahri — CHANGES_REQUESTED** [review]
Also, we need documentation for this, you can find examples in website/ folder

File: opsgenie/data_source_opsgenie_service_test.go (hunk 2 above)
Comment (msg4sunny) [reply opsgenie/data_source_opsgenie_service_test.go]: Just added it to the team, it is not required though.

File: opsgenie/data_source_opsgenie_service.go (hunk 4 above)
Comment (msg4sunny) [inline opsgenie/data_source_opsgenie_service.go:49]: Code didn't got pushed last time, had an issue with IDE

File: opsgenie/data_source_opsgenie_service_test.go (hunk 5 above)
Comment (msg4sunny) [reply opsgenie/data_source_opsgenie_service_test.go:66]: Added team_id and description. Code didn't got pushed last time, had an issue with IDE

File: opsgenie/data_source_opsgenie_service.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+	log.Printf("[INFO] Reading OpsGenie service '%s'", name)
+
+	res, err := client.List(context.Background(), &service.ListRequest{
+		Limit:  10,
```
Comment (ffahri) [inline opsgenie/data_source_opsgenie_service.go]: why do we use limit 10? 
the default value is 20 and max is 100, we can use max value for the avoid request throttling
also why do call this list two different place? we can make this inside for loop below

File: opsgenie/data_source_opsgenie_service_test.go (hunk 2 above)
Comment (ffahri) [reply opsgenie/data_source_opsgenie_service_test.go]: can we delete those since that's not required

**Review by ffahri — CHANGES_REQUESTED** [review]

File: opsgenie/data_source_opsgenie_service.go (hunk 6 above)
Comment (msg4sunny) [reply opsgenie/data_source_opsgenie_service.go]: Changed limit to 100
Calling API once

File: opsgenie/data_source_opsgenie_service_test.go (hunk 2 above)
Comment (msg4sunny) [reply opsgenie/data_source_opsgenie_service_test.go]: Done

File: opsgenie/data_source_opsgenie_service.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+func dataSourceOpsGenieServiceRead(d *schema.ResourceData, meta interface{}) error {
+	client, err := service.NewClient(meta.(*OpsgenieClient).client.Config)
+	if err != nil {
+		return err
+	}
+	name := d.Get("name").(string)
+
+	log.Printf("[INFO] Reading OpsGenie service '%s'", name)
+
+	//offsetString := ""
```
Comment (ffahri) [inline opsgenie/data_source_opsgenie_service.go]: if not used please remove

File: opsgenie/data_source_opsgenie_service.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+			}
+		}
+
+		if res.Paging.Next == "" || breakFlag {
+			break
+		}
+
+		offsetString := strings.Split(res.Paging.Next, string('&'))[2]
+		offsetString = strings.Split(offsetString, string('='))[1]
+		offset, _ = strconv.Atoi(offsetString)
```
Comment (ffahri) [inline opsgenie/data_source_opsgenie_service.go]: please add err check here, if something happens and pagination text may be not correct we dont want to experience crash in the provider . returning error much better than crash

**Review by ffahri — CHANGES_REQUESTED** [review]
Could you add documentation to this implementation, examples can be found in website/ folder in the root of this repository

File: opsgenie/data_source_opsgenie_service.go (hunk 8 above)
Comment (msg4sunny) [reply opsgenie/data_source_opsgenie_service.go]: Added

File: opsgenie/data_source_opsgenie_service.go (hunk 7 above)
Comment (msg4sunny) [reply opsgenie/data_source_opsgenie_service.go]: Done

File: website/docs/d/service.html.markdown [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+# opsgenie\_service
+
+Manages existing Service within Opsgenie.
+
+## Example Usage
+
+```hcl
+data "opsgenie_service" "payment" {
+  name = "payment"
```
Comment (ffahri) [inline website/docs/d/service.html.markdown]: we also need the teamId here

File: website/docs/d/service.html.markdown [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `name` - (Required) Name of the service. This field must not be longer than 100 characters.
```
Comment (ffahri) [inline website/docs/d/service.html.markdown:25]: teamId is also required

File: opsgenie/data_source_opsgenie_service.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+	if err != nil {
+		return err
+	}
+	name := d.Get("name").(string)
+
+	log.Printf("[INFO] Reading OpsGenie service '%s'", name)
+
+	breakFlag := false
+	offset := 0
+	paginationError := errors.New("")
```
Comment (ffahri) [inline opsgenie/data_source_opsgenie_service.go]: we dont need this definition.
we can use 		
```
offset, err = strconv.Atoi(offsetString)
if err!=nil{
return err
}
```
in the 83.line

**Review by ffahri — CHANGES_REQUESTED** [review]
you also need to add document link to here
https://github.com/terraform-providers/terraform-provider-opsgenie/blob/master/website/opsgenie.erb

File: website/docs/d/service.html.markdown (hunk 9 above)
Comment (msg4sunny) [reply website/docs/d/service.html.markdown]: Done

File: website/docs/d/service.html.markdown (hunk 10 above)
Comment (msg4sunny) [reply website/docs/d/service.html.markdown:25]: Done

File: opsgenie/data_source_opsgenie_service.go (hunk 11 above)
Comment (msg4sunny) [reply opsgenie/data_source_opsgenie_service.go]: Changed variable name and removed object creation.

**Comment by msg4sunny** [comment]
> you also need to add document link to here
> https://github.com/terraform-providers/terraform-provider-opsgenie/blob/master/website/opsgenie.erb

Done

**Review by ffahri — APPROVED** [review]

**Review by ffahri — APPROVED** [review]

**Pull request merged**